### ISSUE TITLE
Broken search filter with SQL Mode "ONLY_FULL_GROUP_BY"

### ISSUE THREAD

**Issue by bennyborn** [contao/contao#1661]
**Affected version(s)**

Contao 4.4 and Contao 4.9

**Description**

When searching for e.g. the description of a page in the site structure Contao generates an SQL query which will not work on hosts with MySQL >= 5.7.5 where the SQL mode is set to `ONLY_FULL_GROUP_BY` (which is the default since 5.7.5).

The query in question looks like this

```sql
SELECT DISTINCT `id` FROM tl_page WHERE LOWER(CAST(`description` AS CHAR)) REGEXP LOWER('foo') ORDER BY sorting
```

which results in the following exception

```sql
An exception occurred while executing 'SELECT DISTINCT `id` FROM tl_page WHERE LOWER(CAST(`description` AS CHAR)) REGEXP LOWER('foo') ORDER BY sorting': SQLSTATE[HY000]: General error: 3065 Expression #1 of ORDER BY clause is not in SELECT list, references column 'usr_something.tl_page.sorting' which is not in SELECT list; this is incompatible with DISTINCT
```
Either `sorting` needs to be added to the list of selected fields or omitted entirely.

For reference: We ran into a similar issue in Isotope a couple of years ago (see https://github.com/isotope/core/issues/1886)

**Comment by leofeyer** [comment]
See #1680.

**Comment by fritzmg** [comment]
@leofeyer are you sure that's a duplicate? The issue in #1680 does not affect the core, while the issue of #1661 does affect the core. #1680 is about using the `group` option in Contao Models, while #1661 is about a specific SQL query used by Contao, that is not compatibly with recent SQL versions.

**Comment by bennyborn** [comment]
@fritzmg Just wrote almost the same in #1680 :D

**Comment by m-vo** [comment]
Your originates from here, no?
https://github.com/contao/contao/blob/f90453d0c87cb6674b377fd5cc53fce6ba12e9aa/core-bundle/src/Resources/contao/drivers/DC_Table.php#L5690-L5691

But I don't understand how the `ORDER BY sorting` gets in there. How do you reproduce that in a blank installation?

**Comment by leofeyer** [comment]
I cannot reproduce this in the Contao Official Demo with MariaDB 10.4.12.

This is my sql_mode: `ONLY_FULL_GROUP_BY, STRICT_TRANS_TABLES, ERROR_FOR_DIVISION_BY_ZERO, NO_AUTO_CREATE_USER, NO_ENGINE_SUBSTITUTION`

This is the query: `SELECT id FROM tl_page WHERE LOWER(CAST(`description` AS CHAR)) REGEXP LOWER('b') ORDER BY sorting`

Can anyone else reproduce this? If yes, which version of MySQL/MariaDB and which sql_mode are you using?

**Comment by m-vo** [comment]
@leofeyer What did you enter/select in the backend to get this query? 

(btw.: the query itself does not produce an exception for me either with MariaDB 10.1.44 and `IGNORE_SPACE,ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION` as sql mode)

**Comment by leofeyer** [comment]
Site structure -> search the description of the page -> "b" 😄

**Comment by bennyborn** [comment]
Since the hoster in question has already reacted by removing `ONLY_FULL_GROUP_BY` by default I had to force the SQL mode to trigger the problem.

`SET sql_mode = 'ONLY_FULL_GROUP_BY';`

```
SELECT DISTINCT `id` FROM tl_page WHERE LOWER(CAST(`description` AS CHAR)) REGEXP LOWER('b') ORDER BY sorting;
[Err] 3065 - Expression #1 of ORDER BY clause is not in SELECT list, references column 'xxxxxxx.tl_page.sorting' which is not in SELECT list; this is incompatible with DISTINCT
```

Occured on a MySQL 5.7.28.

Edit: Failed to recreate the problem on a MariaDB so far …

**Comment by m-vo** [comment]
> Site structure -> search the description of the page -> "b" smile

Yeah, but the query that is generated by that does not contain `DISTINCT` like in the example of  @bennyborn?

```
SELECT id FROM tl_page WHERE LOWER(CAST(`description` AS CHAR)) REGEXP LOWER('b') ORDER BY sorting
```

Nevertheless, there is no error with MariaDB even with `DISTICT` in place. Seems like yet another inconsistency. :unamused:

**Comment by leofeyer** [comment]
That's odd. @bennyborn Where does the `SELECT DISTINCT` query come from?

**Comment by fritzmg** [comment]
I also cannot reproduce the query. There are only two places in Contao where `SELECT DISTINCT` is used, neither of which would produce a query like the one @bennyborn mentioned.

**Comment by m-vo** [comment]
Here is some background for why the query with `DISTINCT` fails (in MySQL): A distinct selection over a result set property is internally treated similar to a `GROUP BY` over that property (see https://stackoverflow.com/a/34691355/3146819 for an explanation). 

So in `SELECT DISTICT a FROM table ORDER BY b` the `ORDER BY` would refer to a nonaggregated column or to put it in other words: Ordering would need to happen before the 'distinct filtering' in order to make the latter deterministic. This is not the case, so it fails if `ONLY_FULL_GROUP_BY` is enabled. 

(If this tuns out to be the problem anywhere in the codebase, a solution would be to wrap the order part of the query in a subquery: `SELECT DISTINCT a FROM (SELECT a, b FROM table ORDER BY b)`.)

**Comment by bennyborn** [comment]
Okay, thats strange.

I dug through the Installation where this problem first occured (customer mailed me).
Turns out it was a 4.4.33 and the query was replaced while fixing #269

Before I opened this issue I confirmed I had the same problem in a recent 4.9 installation but so far I had no luck reproducing it again neither did I find anything in DC_Table that could generate such a query.

I guess until I can remember how I got this thing to break in a 4.9 we'll have to call it quits here :\
